OpenRA.Mods.D2k/UtilityCommands/D2kMapImporter.cs

```diff
@@ -0,0 +1,419 @@
+		List<TerrainTemplateInfo> tileSetsFromYaml;
+
+		D2kMapImporter(string filename, string tileset, Ruleset rules)
+		{
+			tilesetName = tileset;
+			this.rules = rules;
+
+			try
+			{
+				stream = File.OpenRead(filename);
+
+				if (stream.Length == 0 || stream.Length % 4 != 0)
+					throw new Exception("The map is in an unrecognized format!");
+
+				Initialize(filename);
+				LoadMapInfo();
+				FillMap();
+			}
+			catch (Exception e)
+			{
+				Console.WriteLine(e);
+				map = null;
+			}
+			finally
+			{
+				stream.Close();
+			}
+		}
+
+		public static Map Import(string filename, string mod, string tileset, Ruleset rules)
+		{
+			var map = new D2kMapImporter(filename, tileset, rules).map;
+			if (map == null)
+				return null;
+
+			map.RequiresMod = mod;
+			map.MakeDefaultPlayers();
+
+			// TODO: Address FixOpenAreas crash
+			// map.FixOpenAreas(rules);
+			return map;
+		}
+
+		void LoadMapInfo()
+		{
+			for (var index = 4; index < stream.Length; index += 4)
+			{
+				var data = stream.ReadUInt16();
+				var specialData = stream.ReadUInt16();
+
+				terrainTiles.Add(GetTile(data));
+				if (specialData != 0)
+					specialValueForTileIndex.Add(index / 4 - 1, specialData);
+			}
+		}
+
+		void Initialize(string mapFile)
+		{
+			mapSize = new Size(stream.ReadUInt16(), stream.ReadUInt16());
+
+			tileSet = rules.TileSets["ARRAKIS"];
+			map = Map.FromTileset(tileSet);
+			map.Title = Path.GetFileNameWithoutExtension(mapFile);
+			map.Author = "Westwood Studios";
+			map.MapSize.X = mapSize.Width + 2 * MapCordonWidth;
+			map.MapSize.Y = mapSize.Height + 2 * MapCordonWidth;
+			map.Bounds = new Rectangle(MapCordonWidth, MapCordonWidth, mapSize.Width, mapSize.Height);
+
+			map.Smudges = Exts.Lazy(() => new List<SmudgeReference>());
+			map.Actors = Exts.Lazy(() => new Dictionary<string, ActorReference>());
+			map.MapResources = Exts.Lazy(() => new CellLayer<ResourceTile>(TileShape.Rectangle, new Size(map.MapSize.X, map.MapSize.Y)));
+			map.MapTiles = Exts.Lazy(() => new CellLayer<TerrainTile>(TileShape.Rectangle, new Size(map.MapSize.X, map.MapSize.Y)));
+
+			map.Options = new MapOptions();
+
+			terrainTiles = new List<TerrainTile>();
+			specialValueForTileIndex = new Dictionary<int, int>();
+
+			// Get all templates from the tileset YAML file that have at least one frame and an Image property corresponding to the requested tileset
+			// Each frame is a tile from the Dune 2000 tileset files, with the Frame ID being the index of the tile in the original file
+			tileSetsFromYaml = tileSet.Templates.Where(t => t.Value.Frames != null
+				&& t.Value.Image.ToLower() == tilesetName.ToLower()).Select(ts => ts.Value).ToList();
+		}
+
+		void FillMap()
+		{
+			for (var y = 0; y < map.MapSize.Y; y++)
+				for (var x = 0; x < map.MapSize.X; x++)
+				{
+					var locationOnMap = new CPos(x, y);
+
+					// Hack for a "clear" tile
+					var indexOnMap = 0;
+
+					if (map.Contains(locationOnMap))
```

Comment:
This is a strange way to parse things.  It would be simpler/clearer to loop scan through the tiles in the source file, and then calculate the equivalent OpenRA map position.